data/nat/basic.lean

```diff
@@ -27,6 +27,46 @@ succ_le_succ_iff
 lemma succ_le_iff {m n : ℕ} : succ m ≤ n ↔ m < n :=
 ⟨lt_of_succ_le, succ_le_of_lt⟩
 
+theorem of_le_succ {n m : ℕ} (H : n ≤ m.succ) : n ≤ m ∨ n = m.succ :=
+(lt_or_eq_of_le H).imp le_of_lt_succ id
+
+@[elab_with_expected_type]
+def le_rec_on {C : ℕ → Sort u} (next : Π {n}, C n → C (n+1)) : Π {n m : ℕ}, n ≤ m → C n → C m
```

Comment:
`next` should have the precondition that `m <= n`

Reply:
I'm afraid that would make `le_rec_on_trans` harder to state (the fact that this is the `Sort` version instead of the `Prop` version means that you would need definitional lemmas).

Reply:
For the Prop version, the precondition n <= n' is really important and natural, while I agree that for the Sort version is it less natural. Resurrecting #555 then.

Reply:
this should be `elab_as_eliminator`

Reply:
I would have thought that the precondition `n <= n'` is really important for the `Sort` version as well. Presumably the main reason for this is when it's not naturally possible to define the function for `n > m`. I don't really see why `Prop` would be any different from `Sort`

Reply:
I have certainly applications in mind with Sort where the precondition is not important (when you have a map from C n to C (n+1) for each n and want to use the inductive limit, for instance), but I agree there are probably examples where it matters.

Reply:
I think the best thing to do is to include `n\le m` and put up with the slightly complicated statement of `le_rec_on_trans`